### PR TITLE
Fix JSON.stringify of an array of nodes

### PR DESCRIPTION
**Pull request by mischnic**
Closes https://github.com/postcss/postcss/issues/1502

I did not know that `JSON.stringify` actually calls `toJSON` on objects...

### CODE REVIEW

File: test/fromJSON.test.ts [hunk 1]
```diff
@@ -37,6 +37,22 @@ it('rehydrates a JSON AST', () => {
 /*# sourceMappingURL=data:application/json;base64,eyJ2ZXJzaW9uIjozLCJzb3VyY2VzIjpbInguY3NzIl0sIm5hbWVzIjpbXSwibWFwcGluZ3MiOiJBQUFzQyxRQUFRLEVBQUUsdUJBQXVCIiwiZmlsZSI6InRvLmNzcyIsInNvdXJjZXNDb250ZW50IjpbIi5mb28geyBjb2xvcjogcmVkOyBmb250LXNpemU6IDEycHQ7IH0gLyogYWJjICovIEBtZWRpYSAod2lkdGg6IDYwZW0pIHsgfSJdfQ== */`)
 })
 
+it('rehydrates an array of Nodes via JSON.stringify', () => {
+  let root = postcss.parse(`
+  .cls {
+    color: orange;
+  }`)
+
+  // @ts-ignore
```
Comment (ai) [inline test/fromJSON.test.ts:46]: Why do need `@ts-ignore`?

File: test/fromJSON.test.ts [hunk 2]
```diff
@@ -37,6 +37,22 @@ it('rehydrates a JSON AST', () => {
 /*# sourceMappingURL=data:application/json;base64,eyJ2ZXJzaW9uIjozLCJzb3VyY2VzIjpbInguY3NzIl0sIm5hbWVzIjpbXSwibWFwcGluZ3MiOiJBQUFzQyxRQUFRLEVBQUUsdUJBQXVCIiwiZmlsZSI6InRvLmNzcyIsInNvdXJjZXNDb250ZW50IjpbIi5mb28geyBjb2xvcjogcmVkOyBmb250LXNpemU6IDEycHQ7IH0gLyogYWJjICovIEBtZWRpYSAod2lkdGg6IDYwZW0pIHsgfSJdfQ== */`)
 })
 
+it('rehydrates an array of Nodes via JSON.stringify', () => {
+  let root = postcss.parse(`
```
Comment (ai) [inline test/fromJSON.test.ts:41]: Can you move to one-line string to keep code style

File: test/fromJSON.test.ts [hunk 3]
```diff
@@ -37,6 +37,22 @@ it('rehydrates a JSON AST', () => {
...
+  }`)
+
+  // @ts-ignore
+  let json = JSON.stringify(root.nodes[0].nodes)
+  let rehydrated = (postcss.fromJSON(
+    JSON.parse(json)
+  ) as unknown) as Declaration[]
+  expect(rehydrated[0].constructor).toBe(postcss.Declaration)
+  // @ts-ignore
+  expect(rehydrated[0].source.input.constructor).toBe(postcss.Input)
```
Comment (ai) [inline test/fromJSON.test.ts:53]:
```suggestion
  expect(rehydrated[0].source.input).toBeInstanceOf(Input)
```

File: lib/node.js [hunk 4]
```diff
@@ -166,7 +166,7 @@ class Node {
     if (!keepBetween) delete this.raws.between
   }
 
-  toJSON (inputs) {
+  toJSON (_, inputs) {
```
Comment (ai) [inline lib/node.js:169]: Why we change arguments?

Comment (mischnic) [reply lib/node.js:169]: JSON.stringify calls toJSON on objects and passes a value as first argument: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify#tojson_behavior

File: test/fromJSON.test.ts (hunk 1 above)
Comment (mischnic) [reply test/fromJSON.test.ts:46]:
```
 FAIL  test/fromJSON.test.ts
  ● Test suite failed to run

    test/fromJSON.test.ts:46:43 - error TS2339: Property 'nodes' does not exist on type 'ChildNode'.
      Property 'nodes' does not exist on type 'Declaration'.

    46   let json = JSON.stringify(root.nodes[0].nodes)
                                                 ~~~~~
    test/fromJSON.test.ts:51:10 - error TS2532: Object is possibly 'undefined'.

    51   expect(rehydrated[0].source.input).toBeInstanceOf(postcss.Input)
                ~~~~~~~~~~~~~~~~~~~~
```

Comment (ai) [reply test/fromJSON.test.ts:43]: Use `rehydrated[0].source?.input` and `rule = root.first as Rule` instead of `@ts-ignore`

**Pull request merged**